CommonTools/RecoAlgos/plugins/PrimaryVertexSorter.h

```diff
@@ -0,0 +1,274 @@
+
+  bool produceOriginalMapping_;
+  bool produceSortedVertices_;
+  bool producePFPileUp_;
+  bool producePFNoPileUp_;
+  int  qualityCut_;
+  bool useMET_;
+};
+
+
+
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "TrackingTools/TransientTrack/interface/TransientTrackBuilder.h"
+#include "TrackingTools/Records/interface/TransientTrackRecord.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+
+// #include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+
+
+using namespace std;
+using namespace edm;
+using namespace reco;
+
+template <class ParticlesCollection>
+PrimaryVertexSorter<ParticlesCollection>::PrimaryVertexSorter(const edm::ParameterSet& iConfig) :
+  assignmentAlgo_(iConfig.getParameterSet("assignment")),
+  sortingAlgo_(iConfig.getParameterSet("sorting")),
+  tokenCandidates_(consumes<ParticlesCollection>(iConfig.getParameter<InputTag>("particles"))),
+  tokenVertices_(consumes<VertexCollection>(iConfig.getParameter<InputTag>("vertices"))),
+  tokenJets_(consumes<edm::View<reco::Candidate> > (iConfig.getParameter<InputTag>("jets"))),
+  produceOriginalMapping_(iConfig.getParameter<bool>("produceAssociationToOriginalVertices")),
+  produceSortedVertices_(iConfig.getParameter<bool>("produceSortedVertices")),
+  producePFPileUp_(iConfig.getParameter<bool>("producePileUpCollection")),
+  producePFNoPileUp_(iConfig.getParameter<bool>("produceNoPileUpCollection")),
+  qualityCut_(iConfig.getParameter<int>("qualityForPrimary")),
+  useMET_(iConfig.getParameter<bool>("usePVMET"))
+{
+
+
+  if(produceOriginalMapping_){
+      produces< CandToVertex> ("original");
+      produces< CandToVertexQuality> ("original");
+  }
+  if(produceSortedVertices_){
+      produces< reco::VertexCollection> ();
+      produces< CandToVertex> ();
+      produces< CandToVertexQuality> ();
+  }
+
+  if(producePFPileUp_){
+      if(produceOriginalMapping_)
+            produces< PFCollection> ("originalPileUp");
+      if(produceSortedVertices_)
+            produces< PFCollection> ("PileUp");
+  }
+
+  if(producePFNoPileUp_){
+      if(produceOriginalMapping_)
+            produces< PFCollection> ("originalNoPileUp");
+      if(produceSortedVertices_)
+            produces< PFCollection> ("NoPileUp");
+  }
+
+
+}
+
+
+
+
+
+template <class ParticlesCollection>
+void PrimaryVertexSorter<ParticlesCollection>::produce(Event& iEvent,  const EventSetup& iSetup) {
+
+
+  Handle<edm::View<reco::Candidate> > jets;
+  iEvent.getByToken( tokenJets_, jets);
+
+  edm::ESHandle<TransientTrackBuilder> builder;
+  iSetup.get<TransientTrackRecord>().get("TransientTrackBuilder", builder);
+
+
+  Handle<VertexCollection> vertices;
+  iEvent.getByToken( tokenVertices_, vertices);
+
+  Handle<ParticlesCollection> particlesHandle;
+  iEvent.getByToken( tokenCandidates_, particlesHandle);
+
+  ParticlesCollection particles = *particlesHandle.product();
```

Comment:
this appears to be pretty cheap for RecoChargedRefCandidatePrimaryVertexSorter  ```std::vector<reco::RecoChargedRefCandidate>``` , likely not so much so for ``` std::vector<reco::PFCandidate>``` in PFCandidatePrimaryVertexSorter